pandas/core/internals/concat.py

```diff
@@ -404,25 +408,29 @@ def _get_empty_dtype_and_na(join_units: Sequence[JoinUnit]) -> Tuple[DtypeObj, A
         else:
             dtypes[i] = unit.dtype
 
+    filtered_dtypes = [
+        unit.dtype for unit in join_units if unit.block is not None and not unit.is_na
+    ]
+    if not len(filtered_dtypes):
+        filtered_dtypes = [unit.dtype for unit in join_units if unit.block is not None]
+    dtype_alt = find_common_type(filtered_dtypes)
+
     upcast_classes = _get_upcast_classes(join_units, dtypes)
 
+    if is_extension_array_dtype(dtype_alt):
```

Comment:
can we just drop the upcast_classes stuff and just do an if/elseif here on dtype_alt? (as you have done with extension / object)

Reply:
isnt't this heavily overlapping with: https://github.com/pandas-dev/pandas/pull/39453 e.g. i like https://github.com/pandas-dev/pandas/pull/39453 as it gets what we want ultimately rigth?

Reply:
i think we're going to get there eventually, but there are a handful of tests that break if we do that right away.  i _think_ the behavior in those tests can be considered buggy (xref #39122), but want to address them in bite-size pieces

Reply:
ok that's fine.